clients/urlclient/interfaces/retry.go

```diff
@@ -0,0 +1,32 @@
+/*******************************************************************************
+ * Copyright 2020 Dell Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ *******************************************************************************/
+
+// interfaces defines contracts specific to the urlclient package
+package interfaces
+
+// RetryStrategy defines some way to verify that a RegistryClient has received data.
+// It accomplishes this by monitoring the values pointed to in its parameters,
+// which communicate the state of the RegistryClient.
+// When the values pointed to by the parameters are in the desired state,
+// the URL that should be passed back to the RegistryClient should be returned by the RetryStrategy.
+type RetryStrategy interface {
+	// Retry defines the actual behavior of the RegistryClient and how it processes the URL.
+	Retry(url *string) (string, error)
+
+	// IsLocked communicates whether the value of the URL is currently being updated or not.
+	IsLocked() bool
+
+	// SetLock updates the value of the lock.
+	SetLock(value bool)
```

Comment:
Are these methods needed for the interface? I see it:

1. A caller shouldn't need invoke either of these methods, but rather only invoke `Retry`
1. The functionality provided are move useful for the implementation to have guards internally, if applicable.
1. Not all implementations will use this. For example, **I think** the `once` implementation could simply be:
```go
// Retry is designed to poll for the data one time. If successful, it returns the data, otherwise it returns an error.
func (o *once) Retry(url *string) (string, error) {
	if url != "" {
		return *url, nil
	}

	return "", errors.NewTimeoutError()
}
```

I bring this up because even though this pattern is useful, the caller still has the burden of managing the locks outside of the `RetryStrategy`:
```go
// newRegistryClient returns a pointer to a registryClient.
// A pointer is used so that when using configuration from a registry, the Prefix can be updated asynchronously.
func newRegistryClient(
	params types.EndpointParams,
	m interfaces.Endpointer,
	strategy urlClientInterfaces.RetryStrategy) *registryClient {

	e := registryClient{
		initialized: false,
		strategy:    strategy,
	}

	go func(ch chan string) {
		for {
			select {
			case url := <-ch:
				e.url = url
				strategy.SetLock(false)
			}
		}
	}(m.Monitor(params))

	return &e
}
``` 

I wonder if we do decide to reduce this down to one method, can this be achieved with type aliasing and higher-order functions. For example:
```go
type RetryStrategy func(*string) error
```

And providing your implementations `once` and `periodic` via higher-order functions which create the `RetryStrategy`. For example:
```go
type RetryStrategy func(*string) error

func NewOnceRetryStrategy() RetryStrategy{
	return func(value *string) error{
		if value != nil {
			return nil
		}
		
		return errors.NewTimeoutError()
	}
}

func NewPeriodicRetryStrategy(timeout time.Duration, ticker time.Duration) RetryStrategy{
	return func(value *string) error{
		if value != nil{
			return nil
		}
		
		// Do waiting logic and return if present
		
		// Else return error
		return errors.NewTimeoutError()
	}	
}
```

This is just a thought, I am not suggesting to do this, just bringing it up as an alternative implementation. I hope all of this makes sense and I have the correct understanding of the implementation.